### PR TITLE
MRG+1: Issue 3593 - Add support for EGI Dig XML File

### PR DESCRIPTION
**Pull request by mattboggess**
I have made edits to read_dig_montage and DigMontage.transform_to_head to support reading EGI digitization xml files. I was hoping someone could do an initial review to see if my edits are appropriate as this is my first time making a contribution.

Still left to do:
- Add test coverage. I have an example dig xml file that could be added to testing data.
- Update whats_new.rst

Closes #3593

### CODE REVIEW

**Comment by mattboggess** [comment]
I added test coverage with the caveat that it currently relies on two raw files and an xml file totaling 3.4 MB being added to the testing data.  Not sure if this is permissible or if an alternative testing approach should be pursued. Also updated whats_new.rst so these changes would now be ready to merge assuming changes are acceptable upon review and testing data can be added.

**Comment by Eric89GXL** [comment]
They should be added to the mne-testing-data repo instead to avoid repo boat

**Comment by mattboggess** [comment]
Ok. I made a pull request to add the new testing files.

**Comment by Eric89GXL** [comment]
Okay, merged and `testing` release is at 0.30 now.

1. Update the version number in `mne/datasets/utils.py` [here](https://github.com/mne-tools/mne-python/blob/master/mne/datasets/utils.py#L228)
2. Either manually download the release and run MD5 on it, or be lazy like me and do `mne.datasets.testing.data_path(force_update=True, verbose=True)` to instead get the new MD5 from the error message.
3. Update the MD5 in `mne/datasets/utils.py` [here])https://github.com/mne-tools/mne-python/blob/master/mne/datasets/utils.py#L278)

**Comment by mattboggess** [comment]
It's throwing a 404 error when trying to download 0.3. Any reason why it wouldn't be able to find the newest version?

**Comment by Eric89GXL** [comment]
Please rebase, too, while you're at it so that the CIs run

File: mne/channels/montage.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 import os.path as op
 
 import numpy as np
+import xml.etree.ElementTree as et
```
Comment (Eric89GXL) [inline mne/channels/montage.py]: let's leave this as `ElementTree` to be explicit. We don't use this anywhere else so when it appears in code `et` will be mysterious

File: mne/channels/montage.py [hunk 2]
```diff
@@ -624,6 +633,44 @@ def read_dig_montage(hsp=None, hpi=None, elp=None, point_names=None,
...
+            if kind == 0:
+                dig_ch_pos['EEG %03d' % number] = coordinates
+            # Reference
+            elif kind == 1:
+                dig_ch_pos['EEG %03d' %
+                           (len(dig_ch_pos.keys()) + 1)] = coordinates
+            # Fiducials
+            elif kind == 2:
+                fid_name = fid_name_map[name]
+                fids[fid_name] = coordinates
```
Comment (Eric89GXL) [inline mne/channels/montage.py]: else warn or error

File: mne/channels/montage.py [hunk 3]
```diff
@@ -624,6 +633,44 @@ def read_dig_montage(hsp=None, hpi=None, elp=None, point_names=None,
...
         elp = np.array(elp) if len(elp) else None
         coord_frame = 'head'
+    elif egi is not None:
+        if not all(x is None for x in (hsp, hpi, elp, point_names, fif)):
+            raise ValueError('hsp, hpi, elp, point_names, fif must all be '
+                             'None if egi is not None')
+        _check_fname(egi, overwrite=True, must_exist=True)
+        if not transform:
+            warn('Dig montage will not be in Neuromag head space. Set '
+                 'transform to True to place in correct coordinate space.')
```
Comment (Eric89GXL) [inline mne/channels/montage.py]: This warning isn't really specific to `egi`. We should probably emit it whenever `coord_frame == 'unknown'`, or not at all

**Comment by mattboggess** [comment]
Ah darn sorry. New to a lot of this. Is there a way to fix the commit history? Or should I just get those changes you suggested made?

**Comment by Eric89GXL** [comment]
`git` has a notoriously challenging learning curve :)

If I were you, I'd do something like this:
```
git fetch upstream
git checkout -b read-egi-gps  # create a new branch
git reset --hard upstream/master  # put it in the state of upstream/master
git cherry-pick 48a3c94 6cdf90b 73d5b11 9e829f8 015a048  # pick relevant commits
```
If this works, then:
```
git push origin --force read-egi-gps:read-egi-gps-dig-file-issue-3953
```
Then carry on with your changes to address my comments, and then
```
git commit -am 'FIX: Address comments'
git push origin read-egi-gps:read-egi-gps-dig-file-issue-3953
```

**Comment by mattboggess** [comment]
Ok I think hopefully that fixed the commit issues and I've also addressed your comments in the latest commit.

File: doc/whats_new.rst [hunk 4]
```diff
@@ -164,6 +167,9 @@ API
 
     - An ``overwrite=False`` default parameter has been added to :func:`write_source_spaces` to protect against accidental overwrites, by `Eric Larson`_
 
+    - An ``egi`` argument has been added to :func:`mne.channels.read_dig_montage` to allow reading in EGI MFF digitization coordinate files. The argument takes the path to the coordinates xml file as a string or defaults to None. by `Matt Boggess`_
```
Comment (Eric89GXL) [inline doc/whats_new.rst]: Technically you don't need this bit. We typically put changes there that could break people's code. Yours is more of an enhancement so the above text suffices. But it's okay to leave it

File: mne/channels/montage.py [hunk 5]
```diff
@@ -624,6 +633,45 @@ def read_dig_montage(hsp=None, hpi=None, elp=None, point_names=None,
...
+            elif kind == 1:
+                dig_ch_pos['EEG %03d' %
+                           (len(dig_ch_pos.keys()) + 1)] = coordinates
+            # Fiducials
+            elif kind == 2:
+                fid_name = fid_name_map[name]
+                fids[fid_name] = coordinates
+            # Unknown
+            else:
+                warn('Unknown sensor type detected. Skipping sensor...'
```
Comment (Eric89GXL) [inline mne/channels/montage.py]: it's good to say `"sensor type (%s) detected..." % (kind,)`

File: mne/channels/montage.py [hunk 6]
```diff
@@ -624,6 +633,45 @@ def read_dig_montage(hsp=None, hpi=None, elp=None, point_names=None,
...
+                     'Proceed with caution!')
+
+        fids = [fids.get(key) for key in ('nasion', 'lpa', 'rpa')]
```
Comment (Eric89GXL) [inline mne/channels/montage.py]: `fids = [fids[key] for key in ...]`

Comment (Eric89GXL) [inline mne/channels/montage.py]: ... or are there times when not all 3 are provided?

**Comment by Eric89GXL** [comment]
FYI by convention we have authors change the title to have `MRG` prefix to signal that they are ready to go from the author's standpoint. I've done that here (with a +1 from me), but feel free to change back to `WIP` if there is more you want to do

**Comment by Eric89GXL** [comment]
I also added the `Closes #3593` to your issue description, which will auto-close the related issue once this is merged

File: doc/whats_new.rst (hunk 4 above)
Comment (mattboggess) [inline doc/whats_new.rst]: deleted

File: mne/channels/montage.py (hunk 5 above)
Comment (mattboggess) [inline mne/channels/montage.py]: Good call. I persisted the type through.

File: mne/channels/montage.py (hunk 6 above)
Comment (mattboggess) [inline mne/channels/montage.py]: I don't know of a case where only a subset may be provided. I can add a check if you think it would be important.

**Comment by mattboggess** [comment]
I addressed your second round of comments and am fine to merge on my end if it all looks good to you and CIs

File: mne/channels/montage.py [hunk 7]
```diff
@@ -624,6 +633,45 @@ def read_dig_montage(hsp=None, hpi=None, elp=None, point_names=None,
...
+            # Fiducials
+            elif kind == 2:
+                fid_name = fid_name_map[name]
+                fids[fid_name] = coordinates
+            # Unknown
+            else:
+                warn('Unknown sensor type %s detected. Skipping sensor...'
+                     'Proceed with caution!' % kind)
+
+        fids = [fids[key] for key in ('nasion', 'lpa', 'rpa')]
```
Comment (Eric89GXL) [inline mne/channels/montage.py]: We can add a check for missing ones later if this fails in an uninformative way. In the meantime we'll assume YAGNI. So this is good as is

**Comment by Eric89GXL** [comment]
Ignore the Travis failure on your previous commit, I killed it so your newest commit can run sooner

**Comment by Eric89GXL** [comment]
Hmm, for some reason this breaks the testing dataset download on OSX. No idea why. I've restarted it a couple of times, same SSL timeout...

In the meantime, anyone else want to have a look?

**Comment by mattboggess** [comment]
It looks like it is working now?

**Comment by Eric89GXL** [comment]
Yep, must be a random download failure issue.

**Comment by mattboggess** [comment]
Is there anything else you need on my end?

**Comment by Eric89GXL** [comment]
Nope, just need another set of eyes to review and merge

**Comment by mattboggess** [comment]
Great, just making sure. Thanks for all of your help!

**Comment by Eric89GXL** [comment]
Thanks @mattboggess !

**Pull request merged**